### ISSUE TITLE
Package-wide parallelism in system tests

### ISSUE THREAD

**Issue by see-quick** [strimzi/strimzi-kafka-operator#5987]
# Package-wide parallelism

## Motivation

In our system tests, we currently do not have package-wide parallelism, which could save a lot of time executing two or more test suites in parallel. Our STs has only method-wide parallelism implemented, which is great but does not provide a significant increase in execution time. In this issue, I want to describe, which problems we currently have with the design of the package-wide parallelism and propose some solutions to how to resolve such problems.

### What we have

1. `Shared Cluster Operator` = deploys only once before all tests are executed in ==BeforeAllOnce.class==. 
2. `@ParallelTest` = annotation, which makes it possible to run simple tests in parallel. By simple, we mean tests that do not contain an additional deployment of Kafka clusters. 
3. `@ParallelNamespaceTest` = annotation, which makes it possible to run more advanced tests in parallel. These tests are basically containing deployments of KafkaMirrorMaker, Kafka cluster and so on...
4. `@IsolatedTest` = annotation, which creates isolation and makes sure that specific test case, which has this kind of annotation run alone without any test case.
5. `@IsolatedSuite` = analogically as @IsolatedTest annotation but this is for isolation of test suite
6. `@ParallelSuite` = analogically as @ParallelTest...

### What  we need

1. We currently don't have resolved RBAC=NAMESPACE problem
2. We don't have re-written all test suites to fully support parallel execution on the class level(except bridge package, plus it means some hours to eliminate race conditions, which could happen when running more Test suites simultaneously) 

### 1 RBAC=NAMESPACE problem

When we deploy Cluster Operator (further only CO), we have to know before all execution, which namespaces will be watched by CO. This brings a problem, where a CO can watch only **n** namespaces where **n** is a non-negative number equal to available CPUs multiplied by 2. So for instance when the machine has 4 CPUs you can only have 8 namespaces (empirically only 7). To solve such a problem we have alternative how to deploy CO with RBAC enabled:

#### 1 alternative

Definitely run these test cases in sequential mode and also for each test suite specify namespaces to watch. Moreover, we must think of the shared CO which, we currently have. In terms that if we always check @BeforeAll in the specific test suite in we have RBAC enabled then Rollback and change configuration of the CO. For instance, could like this code:

```java
class MetricsST extends AbstractST {

	private String namespace1;
	private String namespace2;
	
	@Test
	...
	
	
	@BeforeAll
	void setUp(...) {
		if (RBAC is enabled) {
			// changeConfiguration of shared CO
		} 
		// else we normally execute and also run in parallel if we enabled it...
}

```

#### 2 alternative

The second option is to have a static number of pre-created namespaces, which will be watched by some auxiliary class (i.e., NamespaceWatcher) and each namespace would have a lock and unlock state. For instance:

```plaintext
namespace_1 -> false (can be acquired by any thread)
namespace_2 -> true  (locked by some thread1)
...
namespace_n -> true
```
(where again **n** has to be a non-negative number equal to available CPUs multiplied by 2 )

Moreover, we must consider a parallel approach that would be limited by the **n** number. So here is another proposal to have NamespaceWatcher class only if have RBAC is enabled. This brings another inconsistency to our design, which from MPOV is not a good sign for clean code. 

### 2 RBAC=CLUSTER (implicitly solved)

In this type we have okay because we can use ==\*== to watch all namespaces, which even modify the configuration of the CO. Moreover when one creates new namespaces CO will execute all commands in namespace-wide mode (so that's solved). Another advantage is that we can run tests in parallel very easy and also more Classes at once. The only limitation here is resources, which is not a problem.

### 3 System tests configuration/modes to execute

Assuming that we use the proposed first alternative of the RBAC problem. Then we will get the following matrix, where one can see that we speed up only and only if we use RBAC set to CLUSTER mode, which is in 1/4 case. In other cases, we use sequential execution of system tests.

| platform / install type 	|       OLM      	|      HELM      	|     BUNDLE     	|
|:-----------------------:	|:--------------:	|:--------------:	|:--------------:	|
|    Jenkins (Parallel)   	|  RBAC=CLUSTER  	|  RBAC=CLUSTER  	|  RBAC=CLUSTER  	|
|   Jenkins (Sequential)  	| RBAC=NAMESPACE 	| RBAC=NAMESPACE 	| RBAC=NAMESPACE 	|
|    Azure (Sequential)   	|  RBAC=CLUSTER  	|  RBAC=CLUSTER  	|  RBAC=CLUSTER  	|
|    Azure (Sequential)   	| RBAC=NAMESPACE 	| RBAC=NAMESPACE 	| RBAC=NAMESPACE 	|

## Conclusion

Should we make it? We should decide whether is it okay to increase the runtime of tests, which run with RBAC is set to NAMESPACE and that tradeoff will pay off when we run tests in parallel (RBAC is set to CLUSTER). Moreover, I think that the execution time of overall tests could be decreased by 4 factors (20h -> 5h) in case we have RBAC set to CLUSTER and also set threads in parallel around 10.

**Comment by Frawless** [comment]
Do I understand correctly that RBAC=CLUSTER with package parallelism cut test execution time from 20h to 5h?

I don't think we execute the whole regression with RBAC=NAMESPAE, maybe it would be better to introduce a new tag that will be specific for RBAC=NAMESPACE -> will contain tests that could find the issue and run just only this tag

```	@BeforeAll
	void setUp(...) {
		if (RBAC is enabled) {
			// changeConfiguration of shared CO
		} ``` does not sound like a good solution to me, it would make sense if it could be in `AbstractST` for example

**Comment by see-quick** [comment]
> Do I understand correctly that RBAC=CLUSTER with package parallelism cut test execution time from 20h to 5h?

Hypothetically...but for such parallelism level we need a lot of resources ))

> I don't think we execute the whole regression with RBAC=NAMESPAE, maybe it would be better to introduce a new tag that will be specific for RBAC=NAMESPACE

Yes, that's a good point. But for this subset of tests, I will need to modify the code of the STs.

> does not sound like a good solution to me, it would make sense if it could be in `AbstractST` for example
```
void setUp(...) {
      if (RBAC is enabled) {
      // changeConfiguration of shared CO
      }  
```
This must be in the inner classes because you have to specify namespaces, which are used in that test suite and you do not have such information in the AbstractST.
----
To be clear

Assume we have a scenario:
```
1. MetricsST
    - test1 (tag RBAC)
    - test2
    - test3
2. UserST
    - test1 (tag RBAC)
    - test2
3. TopicST
    - test1 (tag RBAC)
4. KafkaST
    - test1
    - test2 (tag RBAC)
    - test3 (tag RBAC)
    - test4
```
.. other test suites, which does not have RBAC test cases...
---
If we execute only this subset we get:
```
1. MetricsST
    - test1 (tag RBAC)
2. UserST
    - test1 (tag RBAC)
3. TopicST
    - test1 (tag RBAC)
4. KafkaST
    - test2 (tag RBAC)
    - test3 (tag RBAC)
```   

and now in `MetricsST`, we need to modify CO configuration (because we use their 3 namespaces...), so from this configuration
```
// setup cluster operator before all suites only once
install = new SetupClusterOperator.SetupClusterOperatorBuilder()
    .withExtensionContext(sharedExtensionContext)
    .withNamespace(Constants.INFRA_NAMESPACE)
    .withWatchingNamespaces(Constants.WATCH_ALL_NAMESPACES)
    .createInstallation()
    .runInstallation();
```
we need to change to 
```
install = new SetupClusterOperator.SetupClusterOperatorBuilder()
    .withExtensionContext(sharedExtensionContext)
    .withNamespace(Constants.INFRA_NAMESPACE)
    .withWatchingNamespaces("namespace1, namespace2")
    .withBindingNamespaces(ArrayList.asList(namespace1, namespace2))
    .createInstallation()
    .runInstallation();
```
Otherwise, we run all test cases in the `INFRA_NAMESPACE` (its sequential execution so it's okay).

**Comment by see-quick** [comment]
update:

The 1 alternative mentioned above has cons that we have to specify such namespace in inner classes. We may store such info in class that we map test suite names with additional namespaces that are needed. It could looks like this:

```java
public class TestSuiteNamespaceInformer {
    private static final Function<List<String>, List<String>> SUITE_NAMESPACES = (additionalSuiteNamespaces) -> {
        if (NAMESPACED_RBAC) {
            // rbac namespace we are gonna use INFRA_NAMESPACE (everything is in sequential execution)
            return Collections.singletonList(Constants.INFRA_NAMESPACE);
        } else {
            return additionalSuiteNamespaces;
        }
    };

private static TestSuiteNamespaceInformer instance;

    private final Map<String, List<String>> map = Map.of(
        "HttpBridgeKafkaExternalListenersST", SUITE_NAMESPACES.apply(Collections.singletonList(Constants.BRIDGE_KAFKA_EXTERNAL_LISTENER_NAMESPACE)),
        "HttpBridgeCorsST", SUITE_NAMESPACES.apply(Collections.singletonList(Constants.BRIDGE_KAFKA_CORS_NAMESPACE)),
        "HttpBridgeScramShaST", SUITE_NAMESPACES.apply(Collections.singletonList(Constants.BRIDGE_SCRAM_SHA_NAMESPACE)),
        "HttpBridgeTlsST", SUITE_NAMESPACES.apply(Collections.singletonList(Constants.BRIDGE_HTTP_TLS_NAMESPACE)),
     // ... (and more test suites...) 
   );

    public void createNamespace(ExtensionContext extensionContext) {
        final String requiredClassName = extensionContext.getRequiredTestClass().getSimpleName();
        final List<String> namespaces = map.get(requiredClassName);
        final String testSuite = extensionContext.getRequiredTestClass().getName();
        final String testCase = extensionContext.getDisplayName();

        if (namespaces != null) {
            for (String namespaceName : namespaces) {
                KubeClusterResource.getInstance().createNamespace(CollectorElement.createCollectorElement(testSuite, testCase), namespaceName);
            }
        }
    }

    public Map<String, List<String>> getMap() {
        return map;
    }
```
Then In AbstractST, we will have this instance `TestSuiteNamespaceInformer` as a singleton and
 in the `@BeforeAll` method, we will invoke `createNamespace`, which will create namespaces for the specific test suite. Moreover in the specific test suite, we will have instance attribute, which list of namespaces like this:
```java
private final String namespace = testSuiteNamespaceInformer.getMap().get(CruiseControlApiST.class.getSimpleName()).stream().findFirst().get();

```

**Comment by scholzj** [comment]
I think you are over-complicating it:
* The Azure tests are actually running parallelized by using multiple pipelines. Maybe you should consider that as well. It seems to be missing from your considerations.
* The tests should try to follow the expected use-cases and ways the users use the operator. I'm not sure the parallelization follows that.
* Do you count with descoped operators in OLM?

**Comment by see-quick** [comment]
Thanks for the questions :) 

> The Azure tests are actually running parallelized by using multiple pipelines. Maybe you should consider that as well. It seems to be missing from your considerations.

In Azure, we are limited by HW (i.e., `2 CPU 6GM RAM`) and also jobs that we can run simultaneously. Currently (in regression pipeline) we have six jobs, which executes a `sequentially` distinctive subset of whole tests. 

>  Do you count with descoped operators in OLM?

Discussed offline but for the record. Yes, I count with it, currently, we have a shared Cluster Operator, where his configuration for parallel execution will be `cluster-wide` (namespace to watch = `*`).

**Comment by scholzj** [comment]
> > The Azure tests are actually running parallelized by using multiple pipelines. Maybe you should consider that as well. It seems to be missing from your considerations.
>
> In Azure, we are limited by HW (i.e., 2 CPU 6GM RAM) and also jobs that we can run simultaneously. Currently (in regression pipeline) we have six jobs, which executes a sequentially distinctive subset of whole tests.

So, it runs in parallel then if it is 6 jobs, right? ;-) I don't think it in any way disqualifies it as an alternative -> It seems like the most basic and natural way to run the tests in parallel.

**Comment by see-quick** [comment]
Resolved in https://github.com/strimzi/strimzi-kafka-operator/pull/6034
